### ISSUE TITLE
Support for hierarchical data

### ISSUE THREAD

**Issue by McSinyx** [ComLake/core#4]
Or folders in plain English.  AFAICT the following tasks needs to be supported through core's HTTP API:

- Create a new folder from existing data entries (files and folders): this should require metadata fields to `POST /add`.  Create a new entry and return the content ID (CID) to the new folder.
- List entries in a folder: may support recursive walking?  Error on file entries.
- Insert an entry into a folder or subfolder: create new entry with clone of metadata and return CID of new folder.  Error if entry points to a single file.

/cc @larycoder and @npthao1312 for further discussion

**Comment by npthao1312** [comment]
Should both file and folder include source and topics? It feels like a bit of duplication.

**Comment by McSinyx** [comment]
Sorry for the late reply, and yes, that's a good point.  I looked a bit deeper into this and apparently core API should provide a way to add files and directories without/with minimal metadata.  I imagine there are two ways to do this:

1. Having a more versatile `POST /add` which can handle both file and folder (of multiple files and folders) upload.  Implementing this as a multipart file upload is quite simple since I'll only need to proxy the request (with some headers modified) [to IPFS](https://docs.ipfs.io/reference/http/api/#api-v0-add), so it's up to you to decide if you want it (I've no idea how hard it is to implement on the caller side).  Then we'll only need a `POST /insert` (or `PUT /insert` in case of mutable datasets) to cover all user cases.
2. Drop current `POST /add` in favor of lower-level calls, namely `POST /save`, `POST /mkdir`, `POST /cp` (files and folders into a folder) and a new `POST /add` which takes the metadata and the CID (instead of the data).

**Comment by npthao1312** [comment]
Either way is fine for me. But I would prefer 2.

**Comment by npthao1312** [comment]
Remember to update the API doc

**Comment by McSinyx** [comment]
@npthao1312, I've just published a new release with support for this, please refer to the [updated documentation](https://comlake.github.io/comlake.core/api.html) (which is technically still the same URI, relinked for convenience purposes).
